File: scripts/seed_chromadb.py
#!/usr/bin/env python3
"""seed_chromadb.py — Index project docs, knowledge base, and key sources into ChromaDB (Docker).
   Uses HTTP client to connect to ChromaDB running at localhost:8000."""

import chromadb
from chromadb.config import Settings
import os
import re
import sys
from pathlib import Path

BASE = Path(__file__).resolve().parent.parent

# Connect to Docker ChromaDB
client = chromadb.HttpClient(
    host="localhost",
    port=8000,
    settings=Settings(anonymized_telemetry=False)
)

# ── Helper: safe collection name ────────────────────────────────────────────
def safe_name(name: str, max_len: int = 58) -> str:
    cleaned = re.sub(r'[^a-zA-Z0-9_-]', '_', name)
    cleaned = re.sub(r'_{2,}', '_', cleaned)
    cleaned = re.sub(r'^_+', '', cleaned)
    if len(cleaned) <= max_len:
        return cleaned or 'unnamed'
    import hashlib
    h = hashlib.md5(name.encode()).hexdigest()[:8]
    return cleaned[:max_len - 9] + '_' + h

# ── Delete all old collections ──────────────────────────────────────────────
print("Cleaning old collections...")
try:
    for c in client.list_collections():
        try:
            client.delete_collection(c.name)
            print(f"  Deleted: {c.name}")
        except Exception:
            pass
except Exception:
    pass

# ── Chunk text for smaller documents ────────────────────────────────────────
def chunk_text(text: str, max_chars: int = 2000) -> list:
    """Split text into chunks of max_chars, trying to break at paragraph boundaries."""
    if len(text) <= max_chars:
        return [text]
    chunks = []
    paragraphs = text.split('\n\n')
    current = ""
    for p in paragraphs:
        if len(current) + len(p) + 2 <= max_chars:
            current = (current + '\n\n' + p) if current else p
        else:
            if current:
                chunks.append(current.strip())
            current = p if len(p) <= max_chars else p[:max_chars]
    if current:
        chunks.append(current.strip())
    return chunks

# ── Index a group of files into one collection ──────────────────────────────
def index_files(collection_name: str, files: list, doc_type: str):
    """Index all files into a single collection. Each file becomes one document (chunked if large)."""
    if not files:
        return 0

    try:
        collection = client.get_or_create_collection(name=collection_name)
    except Exception as e:
        print(f"  [ERROR] Cannot create collection {collection_name}: {e}")
        return 0

    ids = []
    documents = []
    metadatas = []

    for fpath in files:
        if not fpath.exists():
            continue
        try:
            text = fpath.read_text(encoding="utf-8")
        except Exception:
            print(f"  [WARN] Cannot read: {fpath}")
            continue

        if len(text) < 20:
            continue

        rel = str(fpath.relative_to(BASE))

        # Chunk large files
        chunks = chunk_text(text, max_chars=2000)
        for i, chunk in enumerate(chunks):
            chunk_id = safe_name(f"{rel}__chunk{i}", 63)
            ids.append(chunk_id)
            documents.append(chunk)
            metadatas.append({
                "file": rel,
                "filename": fpath.name,
                "type": doc_type,
                "chunk": i,
                "total_chunks": len(chunks),
                "char_count": len(chunk)
            })

    # Add in batches of 10
    batch_size = 10
    total = 0
    for i in range(0, len(ids), batch_size):
        try:
            collection.add(
                ids=ids[i:i+batch_size],
                documents=documents[i:i+batch_size],
                metadatas=metadatas[i:i+batch_size]
            )
            total += len(ids[i:i+batch_size])
        except Exception as e:
            print(f"  [WARN] Batch add failed at {i}: {e}")

    print(f"  [OK] {collection_name}: {total} docs from {len(files)} files")
    return total

# ═══════════════════════════════════════════════════════════════════════════════
# INDEXING
# ═══════════════════════════════════════════════════════════════════════════════

sys.stdout.reconfigure(encoding='utf-8')
total_docs = 0

# ── 1. Project documentation ────────────────────────────────────────────────
print("\n── Project docs ──")
project_docs = [
    BASE / "ARCHITECTURE.md",
    BASE / "CONVENTIONS.md",
    BASE / "AGENTS.md",
    BASE / "README.md",
    BASE / "CHANGELOG.md",
    BASE / "ЧЕК-ЛИСТ-КОНСОЛИДИРОВАННЫЙ.md",
    BASE / "протокол-сессии.md",
]
total_docs += index_files("project_docs", [f for f in project_docs if f.exists()], "project_doc")

# ── 2. Knowledge base (извлечённые книги) ───────────────────────────────────
print("\n── Knowledge base ──")
kb_dir = BASE / "база-знаний" / "извлечённое"
kb_files = sorted(kb_dir.glob("*.md")) if kb_dir.exists() else []
# Also index prompts
prompts_file = BASE / "база-знаний" / "prompts.md"
if prompts_file.exists():
    kb_files.append(prompts_file)
total_docs += index_files("knowledge_base", kb_files, "knowledge")

# ── 3. UI Kit components ────────────────────────────────────────────────────
print("\n── UI Kit components ──")
ui_files = sorted((BASE / "src" / "app" / "shared" / "ui").glob("kp-*.component.ts"))
# Also index the index.ts barrel file
index_ui = BASE / "src" / "app" / "shared" / "ui" / "index.ts"
if index_ui.exists():
    ui_files.append(index_ui)
total_docs += index_files("source_ui_kit", ui_files, "ui_component")

# ── 4. Core services ────────────────────────────────────────────────────────
print("\n── Core services ──")
core_dir = BASE / "src" / "app" / "core"
core_files = sorted(core_dir.glob("*.ts")) if core_dir.exists() else []
total_docs += index_files("source_core", core_files, "core_service")

# ── 5. Backend ──────────────────────────────────────────────────────────────
print("\n── Backend ──")
backend_files = [
    BASE / "backend" / "src" / "index.ts",
    BASE / "backend" / "src" / "utils" / "crud-factory.ts",
    BASE / "backend" / "src" / "utils" / "crud-factory.spec.ts",
    BASE / "backend" / "src" / "utils" / "api-response.ts",
    BASE / "backend" / "src" / "utils" / "logger.ts",
    BASE / "backend" / "src" / "middleware" / "auth.ts",
    BASE / "backend" / "src" / "middleware" / "error-handler.ts",
    BASE / "backend" / "src" / "config" / "db.ts",
    BASE / "backend" / "src" / "config" / "env.ts",
]
# Add model files
backend_modules = sorted((BASE / "backend" / "src" / "modules").glob("*.ts"))
backend_files.extend(backend_modules)
total_docs += index_files("source_backend", [f for f in backend_files if f.exists()], "backend")

# ── 6. Config & types ───────────────────────────────────────────────────────
print("\n── Config & types ──")
config_files = [
    BASE / "shared" / "types" / "index.ts",
    BASE / "src" / "app" / "app.config.ts",
    BASE / "src" / "app" / "app.routes.ts",
    BASE / "src" / "styles" / "_tokens.scss",
    BASE / "src" / "styles" / "_global.scss",
    BASE / "src" / "app" / "layout" / "admin-layout.component.ts",
]
total_docs += index_files("source_config", [f for f in config_files if f.exists()], "config")

print(f"\n{'='*60}")
print(f"  ChromaDB seeded: {total_docs} documents across all collections")
print(f"  URL: http://localhost:8000")
print(f"{'='*60}")

File: scripts/seed-chromadb.ts
#!/usr/bin/env npx tsx
/**
 * seed-chromadb.ts — Index project docs, knowledge, and source code into ChromaDB (Docker).
 * Usage: npx tsx scripts/seed-chromadb.ts
 *
 * Connects to ChromaDB running in Docker at localhost:8000.
 * Indexes: .md docs, .ts/.html/.scss source files from core/shared/features/layout.
 */

import { ChromaClient } from 'chromadb';
import { readFileSync, existsSync, readdirSync, statSync } from 'fs';
import { extname, join, relative, resolve, basename, dirname } from 'path';
import { fileURLToPath } from 'url';
import { createHash } from 'crypto';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const BASE = resolve(__dirname, '..');

const client = new ChromaClient({ path: 'http://localhost:8000' });

// ── Helpers ────────────────────────────────────────────────────────────────

function safeName(name: string, maxLen = 58): string {
  let cleaned = name.replace(/[^a-zA-Z0-9_-]/g, '_').replace(/_+/g, '_').replace(/^_+/, '');
  if (cleaned.length <= maxLen) return cleaned || 'unnamed';
  const hash = createHash('md5').update(name).digest('hex').slice(0, 8);
  return cleaned.slice(0, maxLen - 9) + '_' + hash;
}

function chunkText(text: string, maxChars = 2000): string[] {
  if (text.length <= maxChars) return [text];
  const chunks: string[] = [];
  const paragraphs = text.split('\n\n');
  let current = '';
  for (const p of paragraphs) {
    if ((current.length + p.length + 2) <= maxChars) {
      current = current ? current + '\n\n' + p : p;
    } else {
      if (current) chunks.push(current.trim());
      current = p.length <= maxChars ? p : p.slice(0, maxChars);
    }
  }
  if (current) chunks.push(current.trim());
  return chunks;
}

function walkFiles(dir: string, extFilter?: string[]): string[] {
  const files: string[] = [];
  if (!existsSync(dir)) return files;

  const walk = (d: string) => {
    for (const entry of readdirSync(d)) {
      const full = join(d, entry);
      const rel = relative(BASE, full);
      // Skip node_modules, .git, dist
      if (rel.includes('node_modules') || rel.includes('.git') || rel.includes('dist') || entry.startsWith('node_modules')) continue;
      try {
        const s = statSync(full);
        if (s.isDirectory()) {
          walk(full);
        } else if (!extFilter || extFilter.includes(extname(entry).toLowerCase())) {
          files.push(full);
        }
      } catch { /* skip */ }
    }
  };
  walk(dir);
  return files;
}

const ALLOWED_SRC_EXTS = ['.ts', '.html', '.scss'];

// ── Index files into a collection ──────────────────────────────────────────

async function indexFiles(
  collectionName: string,
  files: string[],
  docType: string,
): Promise<number> {
  if (files.length === 0) return 0;

  // Delete and recreate collection
  try {
    const existing = await client.listCollections();
    if (existing.find(c => c.name === collectionName)) {
      await client.deleteCollection({ name: collectionName });
    }
  } catch { /* first time */ }

  const collection = await client.createCollection({
    name: collectionName,
    metadata: { 'hnsw:space': 'cosine' },
  });

  const ids: string[] = [];
  const documents: string[] = [];
  const metadatas: Record<string, string>[] = [];

  for (const fpath of files) {
    if (!existsSync(fpath)) continue;
    const rel = relative(BASE, fpath);
    try {
      const text = readFileSync(fpath, 'utf-8');
      if (text.length < 20) continue;

      const chunks = chunkText(text, 2000);
      for (let i = 0; i < chunks.length; i++) {
        ids.push(safeName(`${rel}__chunk${i}`, 63));
        documents.push(chunks[i]);
        metadatas.push({
          file: rel,
          filename: basename(fpath),
          type: docType,
          ext: extname(fpath),
          chunk: String(i),
          total_chunks: String(chunks.length),
          char_count: String(chunks[i].length),
        });
      }
    } catch {
      process.stderr.write(`  [WARN] Cannot read: ${rel}\n`);
    }
  }

  // Add in batches of 20
  const batchSize = 20;
  let total = 0;
  for (let i = 0; i < ids.length; i += batchSize) {
    try {
      await collection.add({
        ids: ids.slice(i, i + batchSize),
        documents: documents.slice(i, i + batchSize),
        metadatas: metadatas.slice(i, i + batchSize),
      });
      total += Math.min(batchSize, ids.length - i);
    } catch (e) {
      process.stderr.write(`  [WARN] Batch add failed at ${i}: ${e}\n`);
    }
  }

  process.stdout.write(`  [OK] ${collectionName}: ${total} docs from ${files.length} files\n`);
  return total;
}

// ═════════════════════════════════════════════════════════════════════════════
// MAIN
// ═════════════════════════════════════════════════════════════════════════════

async function main() {
  process.stdout.write('🧹 Cleaning old collections...\n');
  try {
    const existing = await client.listCollections();
    for (const c of existing) {
      try {
        await client.deleteCollection({ name: c.name });
        process.stdout.write(`  Deleted: ${c.name}\n`);
      } catch { /* ignore */ }
    }
  } catch { /* ignore */ }

  let totalDocs = 0;

  // ── 1. Project documentation ─────────────────────────────────────────────
  process.stdout.write('\n── Project docs ──\n');
  const projectDocs = [
    'ARCHITECTURE.md', 'CONVENTIONS.md', 'AGENTS.md', 'README.md',
    'CHANGELOG.md', 'ЧЕК-ЛИСТ-КОНСОЛИДИРОВАННЫЙ.md', 'протокол-сессии.md',
  ].map(f => join(BASE, f)).filter(existsSync);
  totalDocs += await indexFiles('project_docs', projectDocs, 'project_doc');

  // ── 2. Knowledge base ────────────────────────────────────────────────────
  process.stdout.write('\n── Knowledge base ──\n');
  const kbDir = join(BASE, 'база-знаний', 'извлечённое');
  const kbFiles: string[] = [];
  if (existsSync(kbDir)) {
    for (const f of readdirSync(kbDir)) {
      if (f.endsWith('.md')) kbFiles.push(join(kbDir, f));
    }
  }
  const promptsFile = join(BASE, 'база-знаний', 'prompts.md');
  if (existsSync(promptsFile)) kbFiles.push(promptsFile);
  totalDocs += await indexFiles('knowledge_base', kbFiles, 'knowledge');

  // ── 3. Source code: core services ────────────────────────────────────────
  process.stdout.write('\n── Core services (src/app/core/) ──\n');
  const coreDir = join(BASE, 'src', 'app', 'core');
  const coreFiles = walkFiles(coreDir, ALLOWED_SRC_EXTS);
  totalDocs += await indexFiles('source_core', coreFiles, 'core_service');

  // ── 4. UI Kit components ─────────────────────────────────────────────────
  process.stdout.write('\n── UI Kit (src/app/shared/ui/) ──\n');
  const uiDir = join(BASE, 'src', 'app', 'shared', 'ui');
  const uiFiles = walkFiles(uiDir, ALLOWED_SRC_EXTS);
  totalDocs += await indexFiles('source_ui_kit', uiFiles, 'ui_component');

  // ── 5. Feature pages ─────────────────────────────────────────────────────
  process.stdout.write('\n── Feature pages (src/app/features/) ──\n');
  const featuresDir = join(BASE, 'src', 'app', 'features');
  const featureFiles = walkFiles(featuresDir, ALLOWED_SRC_EXTS);
  totalDocs += await indexFiles('source_features', featureFiles, 'feature');

  // ── 6. Layout ────────────────────────────────────────────────────────────
  process.stdout.write('\n── Layout (src/app/layout/) ──\n');
  const layoutDir = join(BASE, 'src', 'app', 'layout');
  const layoutFiles = walkFiles(layoutDir, ALLOWED_SRC_EXTS);
  totalDocs += await indexFiles('source_layout', layoutFiles, 'layout');

  // ── 7. Styles & tokens ───────────────────────────────────────────────────
  process.stdout.write('\n── Styles (src/styles/) ──\n');
  const stylesDir = join(BASE, 'src', 'styles');
  const styleFiles = walkFiles(stylesDir, ALLOWED_SRC_EXTS);
  totalDocs += await indexFiles('source_styles', styleFiles, 'style');

  // ── 8. Backend ────────────────────────────────────────────────────────────
  process.stdout.write('\n── Backend (backend/src/) ──\n');
  const backendDir = join(BASE, 'backend', 'src');
  const backendFiles = walkFiles(backendDir, ALLOWED_SRC_EXTS);
  totalDocs += await indexFiles('source_backend', backendFiles, 'backend');

  // ── 9. Shared types ──────────────────────────────────────────────────────
  process.stdout.write('\n── Shared types (shared/types/) ──\n');
  const typesDir = join(BASE, 'shared', 'types');
  const typeFiles = walkFiles(typesDir, ALLOWED_SRC_EXTS);
  totalDocs += await indexFiles('source_types', typeFiles, 'config');

  // ── 10. Config (app.config.ts, app.routes.ts, docker-compose) ────────────
  process.stdout.write('\n── Config files ──\n');
  const configFiles = [
    join(BASE, 'src', 'app', 'app.config.ts'),
    join(BASE, 'src', 'app', 'app.routes.ts'),
    join(BASE, 'docker-compose.yml'),
  ].filter(existsSync);
  totalDocs += await indexFiles('source_config', configFiles, 'config');

  // ── Summary ──────────────────────────────────────────────────────────────
  process.stdout.write('\n' + '='.repeat(60) + '\n');
  process.stdout.write(`  ChromaDB seeded: ${totalDocs} documents across all collections\n`);
  process.stdout.write(`  URL: http://localhost:8000\n`);
  process.stdout.write('='.repeat(60) + '\n');
}

main().catch((err) => {
  process.stderr.write(`\n[FATAL] ${err.message}\n`);
  process.stderr.write(`${err.stack}\n`);
  process.exit(1);
});
